### PR TITLE
Symex class cleanup

### PR DESCRIPTION
**Pull request by kroening**

### CODE REVIEW

**Comment by tautschnig** [comment]
tl;dr: I think this is fine, but needs synchronisation with other PRs.

Some remarks:
1. The distinction between `symex_assign` and `symex_assign_rec` was intentional as some cases in "symex_other.cpp" guaranteed that `clean_expr` had already been invoked. I note that one of them didn't, so this actually is a bug fix and it's probably ok to have duplicate calls to `clean_expr`. Well, really that is happening anyway until #1800 is merged.
2. I had tried to help #1641 and #1759 along to enable a merge with the least amount of friction. We'll now have to throw this one in the mix as well. #1827 is a first step on this path, and while technically mergeable right now should receive comments from @kroening @smowton @peterschrammel.

In summary: we've got a lot of pull requests right now, but could do with some communication and reviews.

**Comment by kroening** [comment]
I had a symex_assign_clean initially (for lhs/rhs pairs that are already cleaned), and then noticed it only had one invocation.

**Comment by tautschnig** [comment]
@kroening thanks for the immediate review of #1827. Would you mind rebasing this one so that it can be merged next? @karkhaz and @smowton will then have a bit of work to do to get their respective PRs rebased. (And both have ongoing tasks to factor out individual commits into smaller PRs.)

**Review by tautschnig — APPROVED** [review]
Approving modulo pending rebase and assuming CI will pass.
